massgov/mayflower
# DP-15605 [a11y] sticky toc

Any PRs being created needs a changelog.txt file before being merged into dev. See: [Change Log Instructions](https://github.com/massgov/mayflower/blob/dev/docs/change-log-instructions.md)


## Description

- Set the sticky TOC buttons ("TOP" and "Table of Contents") NOT focusable when they are NOT visible.
- Add context info for screen reader users to the "Show more" button.

Desktop specific:
- Add context info to show the state of the sticky TOC display for screen reader users.
- Add `aria-haspopup` and `aria-expanded` to `button.stickyTOC-open`.
- Replace `<span>` with `<button>` for the overlay close button `.secondary-label-close`.  Also, added `aria-expanded` for the overlay state.
- Remove the pseudo content"x", which gets announced as "plus" by screen readers, added in the markup as invisible to screen readers with its text alternative.
- Set focus on the overlay container `.ma__sticky-toc__column.ma__sticky-toc__stuck-menu.sticky-nav-open` as it opens. So, the content is accessible by keyboard and screen readers.
- Set looping in the focusable elements in the overlay. This prevents focus moves onto elements under the overlay.
- Set focus back on `button.stickyTOC-open` as the overlay closes.

Mobile specific:
- Add `aria-haspopup` and `aria-expanded` to `button.ma__sticky-toc__toggle-link`.
- Remove the pseudo content "+" from `.ma__sticky-toc__toggle-link::after`. Add "+" to `button.ma__sticky-toc__toggle-link` as invisible to screen readers with its text alternative. 
- Set the text alternative to reflect the TOC content display state.


## Related Issue / Ticket

- [JIRA issue](https://jira.mass.gov/browse/DP-15605)

## Steps to Test


1. Go to http://localhost:3000/patterns/05-pages-information-details-table-of-contents/05-pages-information-details-table-of-contents.html.

1. Find there is no visual changes in both desktop and mobile displays whether the TOC content is hidden or shown.

1. Tabbing through the TOC. After the last element of it, no invisible elements get focus. If you try in https://mayflower.digital.mass.gov/?p=pages-information-details-table-of-contents, it takes 2 extra tabbings to get to the "Person to Contact" after the "show more".

1. Open inspector in the browser. 

Desktop:
1. Scroll down below the TOC to display the sticky buttons.

1. Check the markup of the Table of Contents button. It has:
   - `aria-haspopup="true"`
   -  `aria-expanded="false"`
   -  context info for screen reader users as `<span class="ma__visually-hidden">Show</span>`
   ```
   <button class="stickyTOC-open" aria-haspopup="true" aria-expanded="false"><span class="ma__visually-hidden">Show</span>Table of Contents</button>
   ```

1. Click the button to open the overlay.

1. Find the overlay container has focus.  You see the blue border to indicate focus.

1. Hit **TAB** key to navigate the elements inside the container. 
   1. The close button gets focus.  (Originally, any elements in the overlay was not accessible.)
   1. Starting from the close button, going through all TOC elements, then loops back to the close button. It keeps looping while the overlay is open. Focus doesn't go to elements under the overlay.

1. When the close button has focus, hit **ENTER** or **SPACE**.  Focus moves back to the "Table of Contents" button. Then, hitting **TAB** moves focus to the next focusable element in the main page.

1. Check the markup of the close button.  It has:
    - `aria-haspopup="true"`
    - `aria-expanded="false"`
    - visible symbol of "x" is hidden from screen readers with `aria-hidden="true"`
    - text alternative of the symbol as `<span class="ma__visually-hidden">Hide Table of contents</span>`
   ```
   <button class="secondary-label-close" aria-expanded="true"><span class="ma__visually-hidden">Hide Table of contents</span><span aria-hidden="true">x</span></button>
   ```

Mobile:
1. Check the markup of the button.
     <img width="311" alt="Mass_Gov" src="https://user-images.githubusercontent.com/9633303/64464342-9650cd00-d0d5-11e9-92cd-47198204fe37.png">
     - Find `aria-haspopup="true"` and `aria-expanded="false"`.
     - It has context info for screen reader users as `<span class="ma__visually-hidden js-sticky-toc__state">Show</span>` and `<span class="ma__visually-hidden"> table of contents</span>`.
     - The symbol is hidden to screen readers `<span aria-hidden="true">+</span>`.
   ```
    <button class="ma__sticky-toc__toggle-link" aria-haspopup="true" aria-expanded="false"> . 
         <span class="ma__visually-hidden js-sticky-toc__state">Show</span>
         <span class="ma__visually-hidden"> table of contents</span>
         <span aria-hidden="true">+</span>
    </button>
    ```
 1. Click the button and find the following values get updated to reflect the state of the content container.

     Content is close:
      - `aria-expanded="false"`
      - `<span class="ma__visually-hidden js-sticky-toc__state">Show</span>`
      - symbol: "+"


     Content is open:
      - `aria-expanded="true"`
      - `<span class="ma__visually-hidden js-sticky-toc__state">Hide</span>`
      - symbol: "x"

## Screenshots
Use something like [licecap](http://www.cockos.com/licecap/) to capture gifs to demonstrate behaviors.


## Additional Notes:

Anything else to add?

#### Impacted Areas in Application


* 

#### @TODO


*

#### Today I learned...
